### PR TITLE
Refactor MTE-193 [v109] Type string instead of paste for M1

### PR DESCRIPTION
**Pull request by clarmso**
Let me type the string instead of pasting it for M1 if possible for `NavigationTest`, `PhotoActionSheetTest` and `ThirdPartySearchTest`.

Note that some tests from `NavigationTest` are disabled entirely for M1 because those tests are specifically for the copy and paste link functionality.

https://storage.googleapis.com/mobile-reports/public/firefox-ios-M1/FullFunctionalTests/result_64/build/reports/tests.html
https://mozilla-hub.atlassian.net/browse/MTE-183

### CODE REVIEW

File: Tests/XCUITests/FxScreenGraph.swift [hunk 1]
```diff
@@ -640,13 +640,17 @@ func createScreenGraph(for test: XCTestCase, with app: XCUIApplication) -> MMScr
             app.typeText("Mozilla Engine")
             app.tables.textViews["customEngineUrl"].tap()
 
-            UIPasteboard.general.string = "https://developer.mozilla.org/search?q=%s"
-
+            let searchEngineUrl = "https://developer.mozilla.org/search?q=%s"
```
Comment (clarmso) [inline Tests/XCUITests/FxScreenGraph.swift:643]: This code from `FxScreenGraph` is used in `ThirdPartySearchTest`.

File: Tests/XCUITests/NavigationTest.swift [hunk 2]
```diff
@@ -194,83 +195,95 @@ class NavigationTest: BaseTestCase {
...
-        navigator.goto(NewTabScreen)
-        waitForExistence(app.textFields["url"])
-        app.textFields["url"].press(forDuration: 2)
-
-        app.tables.otherElements[ImageIdentifiers.paste].tap()
-        app.buttons["Go"].tap()
-        waitUntilPageLoad()
-        waitForValueContains(app.textFields["url"], value: website_2["moreLinkLongPressInfo"]!)
+    func testCopyLink() throws {
+        if processIsTranslatedStr() == m1Rosetta {
```
Comment (clarmso) [inline Tests/XCUITests/NavigationTest.swift:199]: This test is testing the copy & paste link, so I disable it entirely for M1.

File: Tests/XCUITests/NavigationTest.swift [hunk 3]
```diff
@@ -194,83 +195,95 @@ class NavigationTest: BaseTestCase {
...
-        // Tap on Select All option and make sure Copy, Cut, Paste, and Look Up are shown
-        app.menuItems["Select All"].tap()
-        waitForExistence(app.menuItems["Copy"])
-        if iPad() {
-            XCTAssertTrue(app.menuItems["Copy"].exists)
-            XCTAssertTrue(app.menuItems["Cut"].exists)
-            XCTAssertTrue(app.menuItems["Look Up"].exists)
-            XCTAssertTrue(app.menuItems["Share…"].exists)
+    func testCopyLinkPrivateMode() throws {
+        if processIsTranslatedStr() == m1Rosetta {
```
Comment (clarmso) [inline Tests/XCUITests/NavigationTest.swift:215]: This test is testing the copy & paste link, so I disable it entirely for M1.

File: Tests/XCUITests/NavigationTest.swift [hunk 4]
```diff
@@ -194,83 +195,95 @@ class NavigationTest: BaseTestCase {
...
-        waitForExistence(app.menuItems["Copy"])
-        if iPad() {
-            XCTAssertTrue(app.menuItems["Copy"].exists)
-            XCTAssertTrue(app.menuItems["Cut"].exists)
-            XCTAssertTrue(app.menuItems["Look Up"].exists)
-            XCTAssertTrue(app.menuItems["Share…"].exists)
-            XCTAssertTrue(app.menuItems["Paste & Go"].exists)
-            XCTAssertTrue(app.menuItems["Paste"].exists)
+    func testLongPressOnAddressBar() throws {
+        if processIsTranslatedStr() == m1Rosetta {
```
Comment (clarmso) [inline Tests/XCUITests/NavigationTest.swift:234]: Should I disable this test for M1 as well?

File: Tests/XCUITests/PhotonActionSheetTest.swift [hunk 5]
```diff
@@ -99,11 +99,20 @@ class PhotonActionSheetTest: BaseTestCase {
...
-            waitForExistence(app.collectionViews.buttons["Copy"], timeout: 10)
+        waitForExistence(app.collectionViews.buttons["Copy"], timeout: 10)
+        // This is not ideal but only way to get the element on iPhone
+        // for iPhone/iPad 14 (iOS 16.0), that would be boundBy: 1
+        // For iPhone 14 (iOS 16.1), that would be boundBy: 2
+        // For iPad (iOS 16.1), that would be boundBy: 1
+        var fennecElement = app.collectionViews.scrollViews.cells.element(boundBy: 2)
+
+        if #available(iOS 16.1, *) {
+            if iPad() {
```
Comment (clarmso) [inline Tests/XCUITests/PhotonActionSheetTest.swift:110]: Note that iOS 16.1 on iPhone has the "iMessage" option, so the `boundBy` value is 2.

File: Tests/XCUITests/ThirdPartySearchTest.swift [hunk 6]
```diff
@@ -105,13 +120,6 @@ class ThirdPartySearchTest: BaseTestCase {
         navigator.performAction(Action.AddCustomSearchEngine)
         waitForExistence(app.buttons["customEngineSaveButton"], timeout: 3)
         app.buttons["customEngineSaveButton"].tap()
-        // Workaround for iOS14 need to wait for those elements and tap again
```
Comment (clarmso) [inline Tests/XCUITests/ThirdPartySearchTest.swift:108]: This workaround is no longer applicable.

File: Tests/XCUITests/ThirdPartySearchTest.swift [hunk 7]
```diff
@@ -120,30 +128,34 @@ class ThirdPartySearchTest: BaseTestCase {
...
+            
+            UIPasteboard.general.string = "http://www.google.com/search?q=&btnI"
+            
+            let tablesQuery = app.tables
+            let customengineurlTextView = tablesQuery.textViews["customEngineUrl"].staticTexts["URL (Replace Query with %s)"]
+            
+            XCTAssertTrue(customengineurlTextView.exists)
+            customengineurlTextView.tap()
+            customengineurlTextView.press(forDuration: 2.0)
+            app.staticTexts["Paste"].tap()
```
Comment (clarmso) [inline Tests/XCUITests/ThirdPartySearchTest.swift:150]: I have not been successful in typing in `customengineurlTextView` like the change in `FxScreenGraph`. The `textView` cannot be focused even after I `tap()` and `sleep()` before checking `hasFocus` or attempting to `typeText`. 😞

Comment (isabelrios) [reply Tests/XCUITests/ThirdPartySearchTest.swift:150]: That name is a bit confusing, the `customengineurlTextView` field is actually a `statictText` view...
`customengineurlTextView = tablesQuery.textViews["customEngineUrl"].staticTexts["URL (Replace Query with %s)"]`
if you want to type into the text field you have to use: 
`tablesQuery.textViews["customEngineUrl"].typeText("foo bar...")`
I have tried that locally and it works, I would give this a try before disabling the full test case.

**Comment by isabelrios** [comment]
The `testDownloadLink` is failing on this [line](https://github.com/mozilla-mobile/firefox-ios/pull/12521/files#diff-ce8f284628eb816d3d30a0bf75f540a95facf5e774b12298f5deb33c9eeb98a0L295) because the link saved in the Download panel view is `reserved` instead of `reserved.html`, the next step is to tap on it to see how the page is loading but this is failing, it does not open a website.
I have checked and this works on `iOS 15.5` but not on `iOS 16`... I think this is another issue with iOS16 not M1, I will check but I'm almost sure this will fail on Intel too. We should add it to our list of issues and let the developers know. 
So for now, I would skip this check, the downloaded item name and clicking on it, for iOS 16.

File: Tests/XCUITests/NavigationTest.swift (hunk 4 above)
Comment (isabelrios) [reply Tests/XCUITests/NavigationTest.swift:234]: No, I don't think so... this test checks the options in the ulr bar when we perform a long press... that should work on M1 too as the issue affecting it is related to pasting... 
It works for me locally if I remove the "look up" option.. I don't see that one on the simulator

**Review by isabelrios — CHANGES_REQUESTED** [review]
Requesting changes based on the comments in the PR...not big changes though

File: Tests/XCUITests/NavigationTest.swift (hunk 4 above)
Comment (clarmso) [reply Tests/XCUITests/NavigationTest.swift:234]: Simulator Screen Shot - iPhone 14 Pro Max (iOS 16 0) - 2022-11-28 at 12 41 02
😞 When the test long press on the awesome bar, the pop-up appears. The screenshot is taken on the iPhone 14 Pro Max iOS 16.0 simulator.
`app.textFields["url"].press(forDuration: 3)`

Comment (isabelrios) [reply Tests/XCUITests/NavigationTest.swift:234]: Oh I see ... yeah, then let's disable for now adding a comment that we get that security pop up introduced on iOS 16

File: Tests/XCUITests/NavigationTest.swift [hunk 8]
```diff
@@ -194,83 +195,101 @@ class NavigationTest: BaseTestCase {
...
+            app.textFields["address"].tap()
+            waitForExistence(app.menuItems["Select All"])
+            XCTAssertTrue(app.menuItems["Select All"].exists)
+            XCTAssertTrue(app.menuItems["Select"].exists)
+
+            // Tap on Select All option and make sure Copy, Cut, Paste, and Look Up are shown
+            app.menuItems["Select All"].tap()
+            waitForExistence(app.menuItems["Copy"])
+            if iPad() {
+                XCTAssertTrue(app.menuItems["Copy"].exists)
```
Comment (clarmso) [inline Tests/XCUITests/NavigationTest.swift:251]: @isabelrios Could you please kindly confirm these menu items on your end? I cannot verify this part on my M1.

Comment (isabelrios) [reply Tests/XCUITests/NavigationTest.swift:251]: This is what I see
Simulator Screen Shot - iPad Air (4th generation) - 2022-11-28 at 19 33 45

Comment (clarmso) [reply Tests/XCUITests/NavigationTest.swift:251]: @isabelrios Thanks. How about on the iPhone? ☎️

File: Tests/XCUITests/ThirdPartySearchTest.swift (hunk 7 above)
Comment (clarmso) [reply Tests/XCUITests/ThirdPartySearchTest.swift:152]: You are right! `typeText` works on `textViews` and not on `staticTexts`. 🚀

Comment (clarmso) [reply Tests/XCUITests/ThirdPartySearchTest.swift:152]: I have re-enabled the test. 🎉

File: Tests/XCUITests/NavigationTest.swift (hunk 8 above)
Comment (isabelrios) [reply Tests/XCUITests/NavigationTest.swift:251]: Simulator Screen Shot - iPhone 14 - 2022-11-28 at 21 22 11
Simulator Screen Shot - iPhone 14 - 2022-11-28 at 21 22 16

File: Tests/XCUITests/NavigationTest.swift [hunk 9]
```diff
@@ -96,6 +96,7 @@ class NavigationTest: BaseTestCase {
         navigator.goto(SettingsScreen)
         // Open FxAccount from settings menu and check the Sign in to Firefox screen
         let signInToFirefoxStaticText = app.tables[AccessibilityIdentifiers.Settings.tableViewController].staticTexts[AccessibilityIdentifiers.Settings.FirefoxAccount.fxaSettingsButton]
+        waitForExistence(signInToFirefoxStaticText)
```
Comment (clarmso) [inline Tests/XCUITests/NavigationTest.swift:99]: This test fails sometimes. This line seems to add some stability.

**Comment by isabelrios** [comment]
This PR is ready to land. There is one test failing but it needs to be clarified if that's because a real bug. So @clarmso please file an issue about the `testDownloadLink` I commented above https://github.com/mozilla-mobile/firefox-ios/pull/12521#issuecomment-1328919539 and depending on what is decided we can leave it failing so that we know there is an issue there or we modify it accordingly.

**Review by isabelrios — APPROVED** [review]
As commented, let's land this and follow up on the testDownloadLink test. Thanks

@cso Looks like we can't land this.. there is a Smoke test failing constantly on Bitrise: `testSharePageWithShareSheetOptions` let's investigate why... it says it fails on L120: `waitForExistence(app.navigationBars["ShareTo.ShareView"], timeout: 10)` but when looking at the screenshots, the app is not even launched... I have re-run the build twice with same result... I think the changes [here](https://github.com/mozilla-mobile/firefox-ios/pull/12521/files#diff-713dd9a250745bde1e0b874c2682faa8ca35f23eeb96e686c19c5253f36fc26dR109) are breaking the test

**Comment by clarmso** [comment]
Oops. Rebase failed. I will cherry pick the changes in a new branch. :(

**Comment by clarmso** [comment]
@isabelrios 🙇🏼 I've cherry-picked the changes in a new PR: https://github.com/mozilla-mobile/firefox-ios/pull/12545

**Pull request closed without merging**